design/restic-backup-and-restore-progress.md

```diff
@@ -0,0 +1,104 @@
+This progress will be read on some interval and the PodVolumeBackup Custom Resource's (CR) status will be updated with this information.
+
+### restic restore progress
+
+The `restic ls` command returns the number of files in a backup, including the size of each file.
+This can be compared with the total size of these files in the volume periodically to calculate the completion percentage of the restore.
+The PodVolumeRestore CR's status will be updated with this information.
+
+## Detailed Design
+
+### restic backup progress
+
+restic added support for [streaming JSON output for the `restic backup` command](https://github.com/restic/restic/pull/1944) in 0.9.5.
+Our current images ship restic 0.9.4, and so the Dockerfile will be updated to pull the new version: https://github.com/heptio/velero/blob/af4b9373fc73047f843cd4bc3648603d780c8b74/Dockerfile-velero#L21.
+With the `--json` flag, `restic backup` outputs single lines of JSON reporting the status of the backup:
+
+```
+{"message_type":"status","percent_done":0,"total_files":1,"total_bytes":21424504832}
+{"message_type":"status","action":"scan_finished","item":"","duration":0.219241873,"data_size":49461329920,"metadata_size":0,"total_files":10}
+{"message_type":"status","percent_done":0,"total_files":10,"total_bytes":49461329920,"current_files":["/file3"]}
+{"message_type":"status","percent_done":0.0003815984736061056,"total_files":10,"total_bytes":49461329920,"bytes_done":18874368,"current_files":["/file1","/file3"]}
+{"message_type":"status","percent_done":0.0011765952936188255,"total_files":10,"total_bytes":49461329920,"bytes_done":58195968,"current_files":["/file1","/file3"]}
+{"message_type":"status","percent_done":0.0019503921984312064,"total_files":10,"total_bytes":49461329920,"bytes_done":96468992,"current_files":["/file1","/file3"]}
+{"message_type":"status","percent_done":0.0028089887640449437,"total_files":10,"total_bytes":49461329920,"bytes_done":138936320,"current_files":["/file1","/file3"]}
+```
+
+The [command factory for backup](https://github.com/heptio/velero/blob/af4b9373fc73047f843cd4bc3648603d780c8b74/pkg/restic/command_factory.go#L37) will be updated to include the `--json` flag.
+The code to run the `restic backup` command (https://github.com/heptio/velero/blob/af4b9373fc73047f843cd4bc3648603d780c8b74/pkg/controller/pod_volume_backup_controller.go#L241) will be changed to include a Goroutine that reads from the command's stdout stream.
+The implementation of this will largely follow [@jmontleon's PoC](https://github.com/fusor/velero/pull/4/files) of this.
+The Goroutine will periodically read the stream (every 10 seconds) and get the last printed status line.
+The line will be decoded as JSON and the `percent_done` property will be read and formatted as a percentage value.
+
+The PodVolumeBackupStatus type will be extended to include a `Progress` field.
```

Comment:
I would probably go with `TotalBytes` and `BytesComplete` fields (names TBD), and then calculate percent complete in the UI code.

Reply:
:+1: I think that's useful information to surface. We can fill the TotalBytes field before starting the restore/backup operation.